packages/core/src/common/message-rpc/rpc-message-encoder.ts

```diff
@@ -84,6 +85,108 @@ export class ResponseError extends Error {
     }
 }
 
+/**
+ * Custom error thrown by the {@link RpcMessageEncoder} if an error occurred during the encoding and the
+ * object could not be written to the given {@link WriteBuffer}
+ */
+export class EncodingError extends Error {
+    constructor(msg: string, public cause?: Error) {
+        super(msg);
+    }
+}
+
+/**
+ * A `RpcMessageDecoder` parses a a binary message received via {@link ReadBuffer} into a {@link RpcMessage}
+ */
+export interface RpcMessageDecoder {
+    parse(buffer: ReadBuffer): RpcMessage;
+}
+
+/**
+ * A `RpcMessageEncoder` writes {@link RpcMessage} objects to a {@link WriteBuffer}. Note that it is
+ * up to clients to commit the message. This allows for multiple messages being
+ * encoded before sending.
+ */
+export interface RpcMessageEncoder {
```

Comment:
I tried to minimize API breaks, that's why we stick with the RpcMessageEncoder API for now.
In the long run, this could probably be refactored and cleaned up.
There is no need for an explicit `RpcMessageEncoder`, we could use a more generic `MessageCodec` instead that can simply can encode objects into a `WriteBuffer` and read them from a `ReadBuffer`. But this can be handled in a follow-up.-